### PR TITLE
Bug 1856261 - reset isTablet when onStop

### PR DESCRIPTION
**Pull request by jackyzy823**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1856261

### CODE REVIEW

**Comment by jackyzy823** [comment]
[First version](https://github.com/mozilla-mobile/firefox-android/pull/3888/commits/05c1bb7aa8f79293c2a25f4b8323969609cda90b) couldn't pass test `WHEN fragment configuration change enables tablet size twice THEN tablet action items are only added once` .

I'm not sure what will happens if addTabletActions invoked by onUpdateToolbarForConfigurationChange again. (like 6 buttons?)

File: fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt [hunk 1]
```diff
@@ -122,7 +122,10 @@ class BrowserFragment : BaseBrowserFragment(), UserInteractionHandler {
 
         browserToolbarView.view.addNavigationAction(leadingAction)
 
-        updateToolbarActions(isTablet = resources.getBoolean(R.bool.tablet))
+        this.isTablet = resources.getBoolean(R.bool.tablet)
+        if (this.isTablet) {
+            addTabletActions(requireContext())
+        }
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt:128]: Sorry drive-by: Do we ever care about the else branch here to remove the tablet actions if it's no longer a tablet? Or is it because this is only ever called on `onViewCreated` that we should expect the toolbar to not have any of the tablet actions?

Comment (jackyzy823) [reply fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt:128]: The second. We should expect the toolbar to not have any of the tablet actions.

**Comment by gabrielluong** [comment]
bors try

File: fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt (hunk 1 above)
Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt:128]: That's actually a good question, I couldn't make the app reach `removeTabletActions()` block. I don't have a real foldable device, that might be the issue. @zmckenney Could you help us understand that block of could, in which case do we expect it to get executed?

**Review by mavduevskiy — COMMENTED** [review]
Hey, jackyzy823, thanks for the PR! I remember seeing the bug, it's really easy to hit it. Thanks for taking a look into it.

The solution you provided works just fine, but there are a small things that make me a bit concerned to approve it right away. 

It removes part of the code flow, the change prevents `removeTabletActions` from ever to be called during that lifecycle call. Now we are not exactly sure if it was ever called, but technically it is possible for fragment parameter to be true while the function argument to be false. 

While it doesn't mean that we need to keep the `removeTabletActions` call, if we were to remove it – we should prove that it's a safe thing to do.

It is also related to lifecycle and tablet/foldable devices, so we would like to be extra careful as testing all the cases is challenging. The previous system was aiming to support a cycle of events, triggering decision making on two brunches every time, which is a good approach.

The actual culprit here is the flag that fragment is holding. The flag is meant to keep track of the fragment state (is it a tablet or not), but the Fragment instance can outlive the view, and in that case [this](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt#315) code stops working correctly. It's expected to be called at least once, but it doesn't happen if the fragment retains true value from the previous view. 

It looks like the proper fix would be just resetting the flag when the view gets destroyed, while your solution would kinda "brute force" that case and through it under the carpet. What do you think?

**Comment by gabrielluong** [comment]
> It looks like the proper fix would be just resetting the flag when the view gets destroyed, while your solution would kinda "brute force" that case and through it under the carpet. What do you think?

I do like the idea here to reset the `this.isTablet` flag (possibly during `onStop`?). I had a similar thought when I looked at this PRs awhile back that I was just a bit nervous about avoiding the else branch in `updateToolbarActions` simply because I did not understand the code well and why it was structured the way it is. 

@mavduevskiy Thank you for taking a deeper look into this and considering the different edge cases even if we don't know yet why it was structured the way it is. This is a great instinct to keep developing.

**Review by mavduevskiy — CHANGES_REQUESTED** [review]
Thanks, @jackyzy823!

Just a small change to cover for potential bug and we can land it!

**Review by mavduevskiy — APPROVED** [review]
thanks, @jackyzy823 
lgtm! shipping

**Comment by willdurand** [comment]
@mergifyio backport release_v120

**Comment by willdurand** [comment]
https://github.com/Mergifyio backport releases_v120
